platforms/web/scripts/i18next/parse-translations.ts

```diff
@@ -0,0 +1,66 @@
+import { existsSync, readdirSync, writeFileSync } from 'node:fs';
+import { join } from 'node:path';
+import { execSync } from 'node:child_process';
+
+import prettier from 'prettier';
+
+const i18nextParserConfig = join(process.cwd(), 'i18next-parser.config.js');
+const platformGlob = join(process.cwd(), 'src/**/*.{ts,tsx}');
+
+type WorkspaceEntry = {
+  location: string;
+  workspaceDependencies: string[];
+};
+
+type WorkspaceInfo = Record<string, WorkspaceEntry>;
+
+function syncTranslations() {
+  const result = execSync('yarn workspaces --json info').toString('utf-8');
+  const info = JSON.parse(result);
+  const data = JSON.parse(info.data) as WorkspaceInfo;
+
+  const dependencies = data['@jwp/ott-web']?.workspaceDependencies || [];
+
+  const globs = dependencies
+    .map((dependency) => data[dependency])
+    .filter((entry) => entry && existsSync(join('../..', entry.location, 'src')))
+    .map((entry) => `'${join(entry.location, 'src/**/*.{ts,tsx}')}'`)
+    .join(' ');
```

Comment:
I guess it's kind of a downside that we will end up including all of the translations for any dependant packages, even if only a single function from a single file is really imported. It'd be kind of cool to use a depedency tree tool or something to get all of the individual files that the platform depends on, but that might be inefficient.

Reply:
It will be less sufficient and probably need some extra dependencies, but it sounds more stable than "just" including all files...

I've tried [dependency-cruiser](https://www.npmjs.com/package/dependency-cruiser) real quick and using the API I can quickly list all files used from an entry file.

```ts
import { type IReporterOutput } from 'dependency-cruiser';

async function run() {
  const { cruise } = await import('dependency-cruiser'); // TS says no
  const ARRAY_OF_FILES_AND_DIRS_TO_CRUISE: string[] = ['platforms/web/src'];
  try {
    const cruiseResult: IReporterOutput = await cruise(ARRAY_OF_FILES_AND_DIRS_TO_CRUISE);

    console.log(cruiseResult.output.modules.slice(0, 5));
  } catch (error) {
    console.log(error);
  }
}

run();
```
This will yield:

```js
[
  {
    source: 'platforms/web/src/App.tsx',
    dependencies: [
      [Object], [Object],
      [Object], [Object],
      [Object], [Object],
      [Object], [Object],
      [Object], [Object]
    ],
    dependents: [ 'platforms/web/src/index.tsx' ],
    orphan: false,
    valid: true
  },
  {
    source: 'platforms/web/src/styles/main.scss',
    followable: false,
    coreModule: false,
    couldNotResolve: false,
    matchesDoNotFollow: false,
    dependencyTypes: [ 'local', 'import' ],
    dependencies: [],
    dependents: [ 'platforms/web/src/App.tsx' ],
    orphan: false,
    valid: true
  },
  {
    source: 'platforms/web/src/containers/Root/Root.tsx',
    dependencies: [
      [Object], [Object],
      [Object], [Object],
      [Object], [Object],
      [Object], [Object],
      [Object], [Object],
      [Object], [Object],
      [Object], [Object]
    ],
    dependents: [ 'platforms/web/src/App.tsx' ],
    orphan: false,
    valid: true
  },
  ...
}
```

The arguments will become really big, though; I could quickly try to format the arguments based on the output.

I also see some "unresolved" entries caused by aliasing that need to be solved:

```js
  {
    source: '#src/hooks/useTrackConfigKeyChange',
    followable: false,
    coreModule: false,
    couldNotResolve: true,
    matchesDoNotFollow: false,
    dependencyTypes: [ 'unknown' ]
  }
```

Reply:
I've played with this idea a bit more and now use `tsc --listFilesOnly` to resolve all files.

Next, I created a script that parses the result and builds a glob with nested globs to reduce the size.

For example: 

```
packages/common/src/{utils/{analytics,common}.ts,stores/AccountStore.ts}
```

However, the glob string length is currently 13336 😅

Reply:
What would happen if we just called i18next parse on each file individually?

Reply:
OK, I've tried running the i18next script using every file in the `tsc --listFilesOnly` output. This seems to fail due to unknown reasons inevitably.

I also tried mimicking some of the i18next-parser CLI and use vinyl-fs directly with the i18next-parser transform:

```ts
import { execSync } from 'child_process';
import { pathToFileURL } from 'node:url';

import vfs from 'vinyl-fs';

async function esConfigLoader(filepath: string) {
  return (await import(pathToFileURL(filepath).toString())).default;
}

async function run() {
  const cwd = process.cwd();
  const output = execSync('tsc --listFilesOnly --project ./platforms/web/tsconfig.json');
  let files = output.toString('utf-8').split('\n');

  // @ts-ignore
  const { transform: Transform } = await import('i18next-parser');

  files = files.map((file) => file.replace(cwd, '').slice(1));
  files = files.filter((file) => !file.includes('node_modules') && file !== '');

  const config = (await import(process.cwd() + '/platforms/web/i18next-parser.config.js')).default;

  let count = 0;

  vfs
    .src(files)
    .pipe(
      new Transform(config)
        .on('reading', function (file) {
          count++;
        })
        .on('error', function (message, region) {
          if (typeof region === 'string') message += ': ' + region.trim();
          console.log('  [error]   ' + message);
        })
        .on('warning', function (message) {
          console.log('  [warning] ' + message);
        })
        .on('finish', function () {
          console.log();
          console.log('  Stats:  ' + count + ' files were parsed');
        }),
    )
    .pipe(vfs.dest(process.cwd()));
}

run();
```

This does work! We do lose some flexibility like having a different config, but that shouldn't matter because this script is owned by the platform. 

So, in turn, this script does the following:

- Dry compiles the platform using the tsconfig.json
- Extract all absolute file paths of every require/import
- Makes all paths relative and removes all paths containing node_modules (workspace deps are symlink resolved)
- Pass all paths to vinyl-fs, which reads the files and streams them through i18next-parser
- Write the output like the i18next-parser CLI does

What do you think? 😄

Reply:
@dbudzins @AntonLantukh do you want me to submit a PR with this proposal?

Reply:
Absolutely, if I understand you right, that sounds great! Workspace support for i18n parsing with tree shaking? :heart_eyes: 

This seems like the kind of thing that could be it's own stand-alone open source project contributed to the community if you so wanted. Awesome work!

Reply:
Do you still think the `parse-translations.ts` script belongs to the web platform, or should I make it generic now like the `sync-translations` script in the workspace root?

```
Usage:
  yarn i18next:parse <platform> [resourcePath] [--debug]

  platform           The platform to parse the translations from

  resourcePath       The path to the resources file with generated namespaces
                     Defaults to `src/i18n/resources.ts`

  --debug            Print additional debug information

Use this script to parse all translations from source files
by reading the tsconfig.json.

Example:
  yarn i18next:parse web
```

Reply:
I actually made a different repo for this and published it on NPM! 

- https://www.npmjs.com/package/i18next-parser-workspaces
- https://github.com/ChristiaanScheermeijer/i18next-parser-workspaces

The latest commit installs this package and changes the 18next package script to: 

```shell
$ i18next-workspaces ./ --resources src/i18n/resources.ts
```

The name is not definitive. Let me know if you have suggestions 😄 

Perhaps I can move the sync script as well and use subcommands instead of the confusing CLI options.

```shell
$ i18next-workspaces parse ./
$ i18next-workspaces resources ./src/i18n/resources.ts --locale en --prettier ./.prettierrc
$ i18next-workspaces sync ./ ../app --dry
```

Reply:
Latest update:

- I added subcommands to the i18next-workspaces CLI
- Add the sync (fill-missing) script to i18next-workspaces

If we want to sync two platforms, we can use:

```shell
npx i18next-workspaces ./platforms/web ./platforms/app
```

Reply:
@dbudzins @AntonLantukh any update? 😄